docs: VM Migration Assistant (#4737)
spectrocloud/librarium: docs/docs-content/vm-management/vm-migration-assistant/migrate-vms-vmo-cluster.md

---
sidebar_label: "Migrate VMs to a VMO cluster"
title: "Migrate VMs to a VMO cluster using the VM Migration Assistant"
description: "Learn how to migrate VMs to Palette VMO using the VM Migration Assistant"
icon: " "
hide_table_of_contents: false
sidebar_position: 20
tags: ["vmo", "vm migration assistant"]

> #toc_max_heading_level: 4

Reviewer: We can adjust the layout so that Migrate VMs is an H2, and Cancel Active Migration

![CleanShot 2024-11-22 at 13 24 08](https://github.com/user-attachments/assets/b1d737b1-55e6-4160-8a8d-094f81173fcf)

    [Additional Configuration - Plan Settings](./additional-configuration.md#plan-settings) for guidance.

### Start Migration Plan

1. Once the **Status** field displays **Ready** in the **Plan details**, click **Start migration** in the top-right
   corner.

2. Click **Start** in the pop-up window.

3. Click on the **Virtual Machines** tab.

4. In the table, view the status of the migration for each VM in the **Pipeline status** column.

   Wait until the **Status** shows as **Successful** before [validating](#validate) the migration.

## Validate

1. Log in to [Palette](https://console.spectrocloud.com).

2. From the left **Main Menu**, select **Clusters**. Then, choose the VMO cluster that you migrated your VMs to. The
   **Overview** tab appears.

3. Select the **Virtual Machines** tab. Then, select your migration namespace from the **Namespace** drop-down Menu.
   Your migrated VMs appear.

4. For each migrated VM, click on the **three-dot Menu** and select **Start**. Your VMs are now ready to use.

   ![Start migrated VM](/migrate-vm-kubevirt-guide/vm-management_create-manage-vm_migrate-vm-kubevirt_start_migrated_vm.webp)

> ## Cancel Running Migration

Reviewer: Vale is right

> ## Cancel Running Migration

Author: ```suggestion
## Cancel Active Migration
```

### Prerequisites

> - A running migration plan.

Author: ```suggestion
- An active migration plan.
```

---
sidebar_label: "Migrate VMs to a VMO cluster"
title: "Migrate VMs to a VMO cluster"
description: "Learn how to migrate VMs to Palette VMO using the VM Migration Assistant"
icon: " "
hide_table_of_contents: false
sidebar_position: 20
tags: ["vmo", "vm migration assistant"]
---

Follow this guide to create source providers and migrate your VMs to your VMO cluster.

## Limitations

> - You can only migrate VMs hosted in VMware vSphere.

Reviewer 2: I'd suggest to have some versions of vSphere that we support. Let me check in the slack channel what we should do for that.

- Only VMs whose operating systems are included under
  [`virt-v2v` supported guest systems](https://libguestfs.org/virt-v2v-support.1.html) can be migrated.

## Prerequisites

- A healthy VMO cluster. Refer to the [Create a VMO Profile](../create-vmo-profile.md) for further guidance.

>   - The VMO cluster must have access to VMware and the VMs you want to migrate.

Reviewer 2: ```suggestion
  - The VMO cluster must have access to the VMware vSphere infrastructure and the VMs you want to migrate.
```

Same comment as in the other page, this is too vague. We need to clearly define what type of access.

>   :::warning

Reviewer 2: This complete warning section already exists in one of the regular VMO page, no? Maybe we could just link to there?

  Additionally, we recommend provisioning volumes with the `ReadWriteMany` access mode to ensure that VMs can be
  [live migrated](https://kubevirt.io/user-guide/compute/live_migration/#limitations).

  ```yaml
  kubeadmconfig:
     preKubeadmCommands:
        # Start containerd with new configuration
        - systemctl daemon-reload
        - systemctl restart containerd
     files:
        - targetPath: /etc/containerd/config.toml
        targetOwner: "root:root"
        targetPermissions: "0644"
        content: |
           ## template: jinja

           # Use config version 2 to enable new configuration fields.
           version = 2

           imports = ["/etc/containerd/conf.d/*.toml"]

           [plugins]
              [plugins."io.containerd.grpc.v1.cri"]
              sandbox_image = "registry.k8s.io/pause:3.9"
              device_ownership_from_security_context = true
              [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
              runtime_type = "io.containerd.runc.v2"
              [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
              SystemdCgroup = true
  ```

  :::

> - A VMware vSphere user account with the necessary permissions to manage the VMs you want to migrate.

Reviewer 2: ```suggestion
- A vCenter user account with the necessary permissions to manage the VMs you want to migrate.
```

The first question our customer will ask is what are the necessary VMware permissions required. Also, we don't need to "manage" the source VMs.

  - Migration can be optionally accelerated by providing credentials for the ESXi hosts where the VMs reside.
- One or more VMs hosted in VMware vSphere. Only VMs whose operating systems are included under
  [`virt-v2v` supported guest systems](https://libguestfs.org/virt-v2v-support.1.html) can be migrated.

  - If you are migrating more than one VM in the same plan, they must all share the same network.

>   - For cold migrations, ensure that VMs operating Windows are shut down at the virtualized OS level.

Reviewer 2: ```suggestion
  - For cold migrations, ensure that VMs operating Windows are shut down at the guest OS level.
```

>   - For warm migrations,

Reviewer 2: That's a link to the VDDK documentation for CBT. Must be replaced by https://knowledge.broadcom.com/external/article/315370/enabling-or-disabling-changed-block-trac.html

    [Changed Block Tracking](https://docs.vmware.com/en/VMware-vSphere/8.0/vsphere-vddk-programming-guide/GUID-7B12E618-7851-4BD3-8E39-819454D8C016.html)
    must be enabled on your VMs.

<!--prettier-ignore-->
- The <VersionedLink text="Virtual Machine Migration Assistant" url="/integrations/packs/?pack=vm-migration-assistant-pack"/> pack must be added to your cluster profile. Refer to [Create a VM Migration Assistant Cluster Profile](./create-vm-migration-assistant-profile.md) for guidance.
  - The VM Migration Assistant service console must be accessible from a web browser.

- We recommend providing a
  [VMware Virtual Disk Development Kit (VDDK) image](https://developer.broadcom.com/sdks/vmware-virtual-disk-development-kit-vddk/latest)
  for the migration. This will significantly speed up the migration.

>   - The VDDK image must be built and uploaded to your image registry before starting the migration.

Reviewer 2: Which registry. This piece with the VDDK images in unclear, even for me. It's required for which component? At source or destination?

Author: I've tried making this clearer now.

    <!--prettier-ignore-->
    <details>
    <summary> Example steps to build and upload VDDK image </summary>

    1. Download the VDDK image from the
       [Broadcom Developer Portal](https://developer.broadcom.com/sdks/vmware-virtual-disk-development-kit-vddk/latest).

    2. Decompress the downloaded image.

       ```shell
       tar -xzf VMware-vix-disklib-<version>.x86_64.tar.gz
       ```

    3. Create a Dockerfile to build the VDDK image.

       ```shell
       cat > Dockerfile <<EOF
       FROM <myregistry/myrepository:tag>
       USER 1001
       COPY vmware-vix-disklib-distrib /vmware-vix-disklib-distrib
       RUN mkdir -p /opt
       ENTRYPOINT ["cp", "-r", "/vmware-vix-disklib-distrib", "/opt"]
       EOF
       ```

       Replace the `<myregistry/myrepository:tag>` with your chosen base image registry/repository (for example: `alpine:latest`).

    4. Build the image.

       ```shell
       docker buildx build --platform linux/amd64 --tag <docker-registry>/vddk:<tag> .
       ```

    5. Push the built image to your image registry.

       ```shell
       docker push <docker-registry>/vddk:<tag>
       ```

    </details>

>   - The migration host must have access to your image registry.

Reviewer 2: What's the migration host?

Author: Hangover from the CLI method. Removing.

         if ($VerboseOutput) {
             Get-Content -Path $LogFile
         }

         Write-Log "Check completed. Log saved to $LogFile."
     ```

  2. Make any changes that you need, but ensure that `metadata.name` is set to `forklift-virt-customize`, and `metadata.namespace` is set to your target VM migration namespace.

     Format rules:

     - Windows scripts must follow the regex `^([0-9]+_win_firstboot(([\w\-]*).ps1))$`.
       - For example, `00_win_firstboot.ps1` is a PowerShell script that runs at boot.
     - Linux scripts must follow the regex `^([0-9]+_linux_(run|firstboot)(([\w\-]*).sh))$`.
       - `firstboot` means run at first boot.
       - `run` means execute after virt-v2v conversion, but before the VM starts.
     - The number at the beginning of the key determines the script execution order.

  3. Ensure that your terminal is [configured to access your VMO cluster](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/) and issue the following command to create the ConfigMap on your cluster. 

     ```shell
     kubectl apply -f configmap.yaml
     ```

  </details>

## Migrate VMs

### Create Source Provider

> 1. Log in to the VM Migration Assistant.

Reviewer 2: How? What's the user they must use?

Author: @cloudmaniac  This is the part I don't know, the auth component is still a mystery to me.

Author: I adjusted this to link to the service console section now, which should make things clearer until we have a proper auth method in place.

2. From the left **Main Menu**, select **Providers for virtualization**.

> 3. In the top-left corner, use the **Namespace** drop-down Menu to select your Kubernetes namespace for the migration.

Reviewer 2: It's unclear what this namespace is used for?

Author: @cloudmaniac When you create a plan or provider, they may not show unless you have selected the correct namespace (or All namespaces).

   | **Endpoint type**               | Select the type of endpoint to configure the connection. Choose **vCenter** if managing multiple hosts through a central server, or **ESXi** if connecting directly to a standalone host.                                                                                                                                                                     |
   | **URL**                         | Your vSphere / ESXi API endpoint for the SDK. You can specify a Full Qualified Domain Name (FQDN) or an IP address. For example, `https://vcenter.mycompany.com/sdk`.                                                                                                                                                                                         |
   | **VDDK init image**             | Provide the registry URL to the VMware Virtual Disk Development Kit (VDDK) image, or select **Skip VMware Virtual Disk Development Kit (VDDK) SDK acceleration, migration may be slow.**. If providing an image, make sure you specify the registry URL without the HTTP scheme `https://` or `http://`. For example, `docker.io/myorganization/vddk:v8.0.3`. |

>    | **Username**                    | Your vSphere / ESXi account username. For example, `user@VSPHERE.LOCAL`.                                                                                                                                                                                                                                                                                      |

Reviewer 2: ```suggestion
   | **Username**                    | Your vSphere / ESXi account username. For example, `user@vsphere.local`.                                                                                                                                                                                                                                                                                      |
```

   | **Plan name**        | A unique name for your migration plan.                                                                                                                                                                                                 | `myMigrationPlan`                          |
   | **Target provider**  | Select the target provider from the drop-down Menu. By default, this will be your host cluster.                                                                                                                                        | `host`                                     |
   | **Target namespace** | Select the target namespace for the VM migration from the drop-down Menu.                                                                                                                                                              | `myVmMigrationNamespace`                   |

>    | **Network map**      | A storage map defines the mapping of source storage domains to target storage classes or datastores, ensuring VM disks are correctly placed in the destination environment. Adjust the mapping, or leave the default mapping in place. | `VM-NETWORK` / `Pod Networking`            |

Reviewer 2: ```suggestion
   | **Network map**      | A storage map defines the mapping of source storage domains to target storage classes or datastores, ensuring VM disks are correctly placed in the destination environment. Adjust the mapping, or leave the default mapping in place. | `VM Network` / `Pod Networking`            |
```

>    | **Storage map**      | A network map defines the mapping of source networks to target networks, ensuring VM network interfaces are correctly connected in the destination environment. Adjust the mapping, or leave the default mapping in place.             | `vsanDatastore1` / `spectro-storage-class` |

Reviewer 2: ```suggestion
   | **Storage map**      | A network map defines the mapping of source networks to target networks, ensuring VM network interfaces are correctly connected in the destination environment. Adjust the mapping, or leave the default mapping in place.             | `vsanDatastore` / `spectro-storage-class` |
```

    | **Preserve static IPs**         | Choose whether to preserve the static IPs of the VMs migrated from vSphere.                                                                                                                                                                                     |
    | **Root device**                 | Choose the root filesystem to convert. By default, the first root device is chosen in multi-boot systems. You can specify a root device (for example, `/dev/sda1`) for multi-boot systems, but if it is not detected as a root device, the migration will fail. |

    If you want to explore all additional plan settings, refer to the
    [Additional Configuration - Plan Settings](./additional-configuration.md#plan-settings) for guidance.

11. The plan details have been successfully validated once the plan status displays as **Ready**.

    ![Plan Ready Status](/vm-management_vm-migration-assistant_migrate-vms-vmo-cluster_plan-ready.webp)

> ### Start Warm Migration Plan

Reviewer 2: IMO, cold should come first in the section order for the moment.
